magento/magento2
# Fix for running bin/magento without proper database connection (removes db connection attempts from DI constructors)

Fix for running bin/magento without proper database connection (removes db connection attempts from DI constructors)

Remove obsolete ResourceModel\Product dependency and reworked DateTime to DateTimeFactory. Both of these caused unnecessary connections to be made to the database when executing bin/magento commands.

### Description (*)
Whenever you execute `bin/magento` (with our without explicit commands), Magento will load all registered Commands and load all their dependencies through DI. Those dependencies will load their own dependencies, and so on and so forth.

I've noticed that through the whole depenencies tree, some dependencies tried to fetch something from the database. This is really not needed when executing `bin/magento` without providing an explicit command to actually do something with the database.

The removal of the `Magento\Catalog\Model\ResourceModel\Product` dependency should be safe, as it wasn't being used in the class anyways.
The change of the `DateTime` to `DateTimeFactory` causes that no locale info from the database is being requested in the constructor. It should only happen at runtime and [was inspired by this code here](https://github.com/magento/magento2/blob/2.4.6/app/code/Magento//MediaContentSynchronization/Model/Synchronize.php#L58).

These 2 dependencies were loaded from the commands `bin/magento media-gallery:sync`  & `bin/magento media-content:sync`

### Related Pull Requests

### Fixed Issues (if relevant)
1. Fixes https://github.com/magento/magento2/issues/37409
2. Fixes https://github.com/magento/magento2/issues/36845

### Manual testing scenarios (*)
1. Setup clean Magento and install it
3. Modify the `app/etc/env.php` file and change the correct database user to an incorrect one
4. Run `bin/magento`
5. Expected is that no error messages show up until you actually want to do something to the database with an explicit command
6. Change back to a correct database user in the `app/etc/env.php` file
7. Make sure both the commands `bin/magento media-gallery:sync` & `bin/magento media-content:sync` still work as expected, those should be the only real commands affected by these changes

### Questions or comments

### Contribution checklist (*)
